src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java

```diff
@@ -673,43 +631,29 @@ private boolean shouldInstantiate(TypeElement site, Element e) {
      * superclass is java.lang.Object no matter how many other interfaces
      * that interface extends.
      */
-    public DeclaredType overriddenType(ExecutableElement method) {
-        return configuration.workArounds.overriddenType(method);
-    }
-
-    private  TypeMirror getType(TypeMirror t) {
-        return (isNoType(t)) ? getObjectType() : t;
-    }
-
-    public TypeMirror getSuperType(TypeElement te) {
-        TypeMirror t = te.getSuperclass();
-        return getType(t);
-    }
-
-    public ExecutableElement overriddenMethod(ExecutableElement method) {
-        if (isStatic(method)) {
-            return null;
-        }
-        final TypeElement origin = getEnclosingTypeElement(method);
-        for (TypeMirror t = getSuperType(origin);
-             t.getKind() == DECLARED;
-             t = getSuperType(asTypeElement(t))) {
-            TypeElement te = asTypeElement(t);
-            if (te == null) {
+    public OverrideInfo overriddenMethod(ExecutableElement method) {
+        var t = method.getEnclosingElement().asType();
+        // in this context, consider java.lang.Object to be the superclass of an interface
+        while (true) {
+            var supertypes = typeUtils.directSupertypes(t);
+            if (supertypes.isEmpty()) {
+                // reached the top of the hierarchy
+                assert typeUtils.isSameType(getObjectType(), t);
                 return null;
             }
+            t = supertypes.get(0);
+            // if non-empty, the first element is always the superclass
+            var te = (TypeElement) ((DeclaredType) t).asElement();
+            assert te.getKind().isClass();
             VisibleMemberTable vmt = configuration.getVisibleMemberTable(te);
             for (Element e : vmt.getMembers(VisibleMemberTable.Kind.METHODS)) {
```

Comment:
In similar situations, I have found it useful to pass in a type token that determines the result type and which avoids casting the returned objects:

For example:
```
   <T> List<T extends Element> getMembers(VisibleMemberTable.Kind k, Class<T> clazz)
```

Reply:
Unfortunately, there's no such thing as generic enum, which would be just the ticket here.

Type token would be okay, but perhaps only marginally better than what we have now: neither option supports static checking. Anyhow, we might revisit it in the next round of refactoring, not now.

Reply:
OK, with not now, but I do think the type token is marginally better than a cast of the result at the use site.